### PR TITLE
Change underscore dependency to lodash

### PR DESCRIPTION
**Pull request by jpchen**
Addresses #530. Will need an `npm install` to work locally.

Tests:
Passes all existing tests.  

Ran into some merging issues and couldn't squash locally, feel free to squash before merging.

### CODE REVIEW

File: src/header.wppl [hunk 1]
```diff
@@ -531,7 +531,7 @@ var Vector = function(arr) {
 var Matrix = function(arr) {
   var n = arr.length;
   var m = arr[0].length;
-  var t = ad.tensor.fromScalars(_.flatten(arr));
+  var t = ad.tensor.fromScalars(_.flattenDeep(arr));
```
Comment (stuhlmueller) [inline src/header.wppl]: Do we only need `_.flatten` here? I think the assumption is that the input array is a list of lists, e.g. `[[1, 2], [3, 4]]`.

File: src/util.js [hunk 2]
```diff
@@ -271,7 +271,7 @@ function timeif(bool, name, thunk) {
 }
 
 function pipeline(fns) {
-  return _.compose.apply(null, fns.reverse());
+  return _.flowRight.apply(null, fns.reverse());
```
Comment (stuhlmueller) [inline src/util.js]: Would just `_.flow(fns)` be equivalent to this?

(I'm a little surprised to see that the current `pipeline` is mutating the `fns` argument, but I suppose `pipeline` is only used on arrays that are created for the sole purpose of being passed to it.)

Comment (jpchen) [inline src/util.js]: `_.flow` invokes functions left to right, whereas `_.compose`and `_.flowRight` is a normal function composition you'd expect where the last argument is the innermost fn, etc. A bit confusing, take a look here: https://lodash.com/docs/4.17.2#flow

File: src/header.wppl (hunk 1 above)
Comment (jpchen) [inline src/header.wppl]: This was to be paranoid/safe when migrating. `flatten` in underscore is deep unless a shallow argument passed, while lodash breaks them up into two functions, `flatten` which is shallow and `flattenDeep` which is deep.  I guess if the matrix is always 2D, I can change it.

Comment (stuhlmueller) [inline src/header.wppl]: I think the matrix is always 2D, and that something else has gone wrong if it's not. (Paul, correct me if I'm wrong.) In that case, flattening only one step could make errors easier to find since `fromScalars` will hopefully throw an error if it gets a list as input.

File: src/util.js (hunk 2 above)
Comment (stuhlmueller) [inline src/util.js]: But leaving out `reverse` as suggested would recover the intended order, no?

**Comment by null-a** [comment]
Am I correct in thinking this will break any webppl programs that rely on functions present in underscore but not in lodash? (e.g. `_.object`.)

**Comment by stuhlmueller** [comment]
Yes, such programs will have to be updated. I don't think this will be very common, the upgrade is easy, and the next release will include a number of breaking changes anyway. We can point people to the [underscore->lodash migration page](https://github.com/lodash/lodash/wiki/Migrating) in our release notes.

File: src/header.wppl (hunk 1 above)
Comment (jpchen) [inline src/header.wppl]: Done.

File: src/util.js (hunk 2 above)
Comment (jpchen) [inline src/util.js]: Ah yes my apologies, I misread your comment. Fixed.

**Comment by jpchen** [comment]
Yes, I also hastily whipped up cross documentation between the two libraries, for ease of future use: jonathanpchen.com/underdash-api
You can look up the equivalent lodash and underscore methods.

It's incomplete and based off of what I used above and other functions I looked into, but I plan to update as I go.

File: src/util.js [hunk 3]
```diff
@@ -271,7 +271,7 @@ function timeif(bool, name, thunk) {
 }
 
 function pipeline(fns) {
-  return _.flowRight.apply(null, fns.reverse());
+  return _.flow.apply(null, fns);
```
Comment (stuhlmueller) [inline src/util.js]: I think `_.flow(fns)` would be equivalent for our purposes, since `_.flow` can take arrays as well. But at that point, it's not clear that we even need the `pipeline` function, since `_.flow` is fewer characters than `util.pipeline`.

File: scripts/distributionDocs [hunk 4]
```diff
@@ -1,14 +1,14 @@
-#!/usr/bin/env node
+#!usr/bin/env node
```
Comment (null-a) [inline scripts/distributionDocs]: Is this change intentional?

File: src/util.js (hunk 3 above)
Comment (null-a) [inline src/util.js]: I added `pipeline` because there wasn't an equivalent in underscore, but yeah, sounds like we can get rid of it now. (At some point.)

File: scripts/distributionDocs (hunk 4 above)
Comment (jpchen) [inline scripts/distributionDocs]: Fixed. Must've been something in my bash settings when I ran build...

File: src/util.js (hunk 3 above)
Comment (jpchen) [inline src/util.js]: Done.
